sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +376,368 @@ def transform(self, X):
             X[coordinates] = values
 
         return X
+
+
+class KNNImputer(BaseEstimator, TransformerMixin):
+    """Imputation for completing missing values using k-Nearest Neighbors.
+
+    Each sample's missing values are imputed from up to ``max_neighbors``
+    nearest neighbors found in the training set. Each missing feature is then
+    imputed as the average, either weighted or unweighted, of these neighbors
+    who have a value for it. Where all neighbors have that feature value
+    missing, the training set average for that feature is used for imputation.
+
+    Parameters
+    ----------
+    missing_values : integer or "NaN", optional (default = "NaN")
+        The placeholder for the missing values. All occurrences of
+        `missing_values` will be imputed. For missing values encoded as np.nan,
+        use the string value "NaN".
+
+    max_neighbors : int, optional (default = 5)
+        Maximum number of neighboring samples to use for imputation. When any
+        of the neighbors themselves have the feature value missing then the
+        remaining neighbors, if any, that have the feature value available are
+        used. But if none of the neighbors have the value available, the global
+        feature mean (i.e., by default, the column mean) is used for
+        imputation.
+
+    weights : str or callable, optional (default = "uniform")
+        Weight function used in prediction.  Possible values:
+
+        - 'uniform' : uniform weights.  All points in each neighborhood
+          are weighted equally.
+        - 'distance' : weight points by the inverse of their distance.
+          in this case, closer neighbors of a query point will have a
+          greater influence than neighbors which are further away.
+        - [callable] : a user-defined function which accepts an
+          array of distances, and returns an array of the same shape
+          containing the weights.
+
+    metric : str or callable, optional (default = "masked_euclidean")
+        Distance metric for searching neighbors. Possible values:
+        - 'masked_euclidean'
+        - [callable] : a user-defined function which conforms to the
+        definition of _pairwise_callable(X, Y, metric, **kwds). In other
+        words, the function accepts two arrays, X and Y, and a
+        ``missing_values`` keyword in **kwds and returns a scalar distance
+        value.
+
+    row_max_missing : float, optional (default = 0.5)
+        The maximum percentage of columns (i.e. features) that can be missing
+        before the sample is excluded from nearest neighbor imputation. It
+        means that such rows will not be considered a potential donor in fit(),
```

Comment:
donor seems odd. Also double backticks on functions. Where does the default come from?

Reply:
default from R package